File: src/main/java/AoC2017_15.java
import static com.github.pareronia.aoc.IntegerSequence.Range.range;

import java.util.Iterator;
import java.util.List;
import java.util.function.Predicate;

import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2017_15
        extends SolutionBase<AoC2017_15.Generators, Integer, Integer> {
    
    private static final long FACTOR_A = 16807;
    private static final long FACTOR_B = 48271;
    private static final long MOD = 2147483647;
    private static final Predicate<Long> NONE = x -> true;

    private AoC2017_15(final boolean debug) {
        super(debug);
    }

    public static AoC2017_15 create() {
        return new AoC2017_15(false);
    }

    public static AoC2017_15 createDebug() {
        return new AoC2017_15(true);
    }

    @Override
    protected AoC2017_15.Generators parseInput(final List<String> inputs) {
        return new Generators(
                Long.parseLong(inputs.get(0).split(" ")[4]),
                Long.parseLong(inputs.get(1).split(" ")[4]));
    }

    private Integer solve(
            final Generators generators,
            final int reps,
            final Predicate<Long> conditionA,
            final Predicate<Long> conditionB
    ) {
        final Iterator<Long> iteratorA = generators.iteratorA(conditionA);
        final Iterator<Long> iteratorB = generators.iteratorB(conditionB);
        return (int) range(reps).intStream()
            .filter(i -> iteratorA.next().shortValue() == iteratorB.next().shortValue())
            .count();
    }
    
    private Predicate<Long> isMultipleOf(final int d) {
        return x -> x % d == 0;
    }
    
    @Override
    public Integer solvePart1(final Generators generators) {
        return solve(generators, 40_000_000, NONE, NONE);
    }
    
    @Override
    public Integer solvePart2(final Generators generators) {
        return solve(generators, 5_000_000, isMultipleOf(4), isMultipleOf(8));
    }

    @Samples({
        @Sample(method = "part1", input = TEST, expected = "588"),
        @Sample(method = "part2", input = TEST, expected = "309"),
    })
    public static void main(final String[] args) throws Exception {
        AoC2017_15.create().run();
    }
    
    private static final String TEST = """
            Generator A starts with 65
            Generator B starts with 8921
            """;
    
    record Generators(long a, long b) {

        public Iterator<Long> iteratorA(final Predicate<Long> condition) {
            return createIterator(a, FACTOR_A, condition);
        }

        public Iterator<Long> iteratorB(final Predicate<Long> condition) {
            return createIterator(b, FACTOR_B, condition);
        }

        private Iterator<Long> createIterator(
                final long seed,
                final long factor,
                final Predicate<Long> condition
        ) {
            return new Iterator<>() {
                private long prev = seed;
                
                @Override
                public Long next() {
                    long val = prev;
                    while (true) {
                        val = (val * factor) % MOD;
                        if (condition.test(val)) {
                            prev = val;
                            return val;
                        }
                    }
                }
                
                @Override
                public boolean hasNext() {
                    return true;
                }
            };
        }
    }
}

File: src/main/python/AoC2017_15.py
#! /usr/bin/env python3
#
# Advent of Code 2017 Day 15
#

from __future__ import annotations

import sys
from typing import Callable
from typing import NamedTuple

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

FACTOR_A = 16807
FACTOR_B = 48271
MOD = 2147483647

TEST = """\
Generator A starts with 65
Generator B starts with 8921
"""


class Generator:
    def __init__(
        self, seed: int, factor: int, condition: Callable[[int], bool]
    ) -> None:
        self.prev = seed
        self.factor = factor
        self.condition = condition

    def __iter__(self) -> Generator:
        return self

    def __next__(self) -> int:
        val = self.prev
        while True:
            val = (val * self.factor) % MOD
            if self.condition(val):
                self.prev = val
                return val


class Generators(NamedTuple):
    a: int
    b: int

    @classmethod
    def from_input(cls, input: list[str]) -> Generators:
        return Generators(*(int(input[i].split()[-1]) for i in (0, 1)))

    def generator_a(self, condition: Callable[[int], bool]) -> Generator:
        return Generator(self.a, FACTOR_A, condition)

    def generator_b(self, condition: Callable[[int], bool]) -> Generator:
        return Generator(self.b, FACTOR_B, condition)


Input = Generators
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return Generators.from_input(list(input_data))

    def solve(
        self,
        generators: Generators,
        reps: int,
        condition_a: Callable[[int], bool],
        condition_b: Callable[[int], bool],
    ) -> int:
        generator_a = generators.generator_a(condition_a)
        generator_b = generators.generator_b(condition_b)
        return sum(
            next(generator_a) & 0xFFFF == next(generator_b) & 0xFFFF
            for _ in range(reps)
        )

    def part_1(self, generators: Generators) -> int:
        return self.solve(
            generators, 40_000_000, lambda x: True, lambda x: True
        )

    def part_2(self, generators: Generators) -> int:
        return self.solve(
            generators, 5_000_000, lambda x: x % 4 == 0, lambda x: x % 8 == 0
        )

    @aoc_samples(
        (
            ("part_1", TEST, 588),
            ("part_2", TEST, 309),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2017, 15)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
